aptos-move/framework/aptos-framework/sources/object.move

```diff
@@ -0,0 +1,433 @@
+        };
+
+        event::emit_event(
+            &mut object.transfer_events,
+            TransferEvent {
+                object_id: object_id,
+                from: object.owner,
+                to,
+            },
+        );
+        object.owner = to;
+    }
+
+    public fun transfer_to_object(
+        owner: &signer,
+        object_id: ObjectId,
+        to: ObjectId,
+    ) acquires Object {
+        transfer(owner, object_id, to.inner)
+    }
+
+    /// This checks that the destination address is eventually owned by the owner and that each
+    /// object between the two allows for ungated transfers.
+    fun verify_ungated_and_descendant(owner: address, destination: address) acquires Object {
+        let current_address = destination;
+        assert!(
+            exists<Object>(current_address),
+            error::not_found(EOBJECT_DOES_NOT_EXIST),
+        );
+
+        while (owner != current_address) {
+            // At this point, the first object exists and so the more likely case is that the
+            // object's owner is not an object. So we return a more sensible error.
+            assert!(
+                exists<Object>(current_address),
+                error::permission_denied(ENOT_OBJECT_OWNER),
+            );
+            let object = borrow_global<Object>(current_address);
+            assert!(
+                object.allow_ungated_transfer,
+                error::permission_denied(ENO_UNGATED_TRANSFERS),
+            );
+
+            current_address = object.owner;
+        };
+    }
+
+    /// Accessors
+
+    public fun owner(object_id: ObjectId): address acquires Object {
+        assert!(
+            exists<Object>(object_id.inner),
+            error::not_found(EOBJECT_DOES_NOT_EXIST),
+        );
+        borrow_global<Object>(object_id.inner).owner
+    }
+
+    public fun is_owner(object_id: ObjectId, owner: address): bool acquires Object {
+        owner(object_id) == owner
+    }
+
+    native fun create_signer(addr: address): signer;
+
+    #[test_only]
+    use std::option::{Self, Option};
+
+    #[test_only]
+    const EHERO_DOES_NOT_EXIST: u64 = 0x100;
+    #[test_only]
+    const EWEAPON_DOES_NOT_EXIST: u64 = 0x101;
+
+    #[test_only]
+    struct HeroEquipEvent has drop, store {
+        weapon_id: Option<ObjectId>,
+    }
+
+    #[test_only]
+    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+    struct Hero has key {
+        equip_events: event::EventHandle<HeroEquipEvent>,
+        weapon: Option<ObjectId>,
+    }
+
+    #[test_only]
+    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+    struct Weapon has key { }
+
+    #[test_only]
+    public fun create_hero(creator: &signer): ObjectId acquires Object {
+        let hero_creator_ref = create_named_object(creator, b"hero");
+        let hero_signer = generate_signer(&hero_creator_ref);
+        let guid_for_equip_events = create_guid(&hero_signer);
+        move_to(
```

Comment:
I can also use the hero_signer to move a weapon resource to the account. In this case, the one object_id can have both weapon and hero?

Reply:
Yeah, there's no restrictions on how you want to configure your object / resources -- at least not the way that move works today... we'd have to introduce a new native function to obtain that kind of functionality -- or we could attempt to tweak the VM

Reply:
yea, ideally, the ref itself can specify to object it can create and modify